\documentclass{beamer}
\usetheme{metropolis}
\usepackage[utf8]{inputenc}
\usepackage{amsfonts}
\usepackage{amsmath}
\usepackage{natbib}
\usepackage{graphicx}
\usepackage{array,booktabs,tabularx}
\usepackage{epstopdf}
\usepackage{colortbl, xcolor}
\usepackage{url}

\setlength\abovecaptionskip{-8pt}
\setbeamertemplate{caption}{\raggedright\insertcaption\par}

\title{n-gram analysis of biological sequences in R}
\date{}
\author{Micha\l{} Burdukiewicz\inst{1}, Piotr Sobczyk\inst{2}, Ma\l{}gorzata Kotulska\inst{3}, Pawe\l{} Mackiewicz\inst{1}}
\institute{\textsuperscript{1}University of Wroc\l{}aw, Department of Genomics, 

\textsuperscript{2}Wroc\l{}aw University of Science and Technology, Faculty of Pure and Applied Mathematics,

\textsuperscript{3}Wroc\l{}aw University of Science and Technology, Department of Biomedical Engineering
}

\begin{document}
  \maketitle
  \section{Introduction}
  
\begin{frame}{Biological sequences}
<<knitrIntro, echo = FALSE, message=FALSE,warning=FALSE>>=
source("my_ggplot_theme.R")

library(biogram)
library(seqinr)
library(dplyr)
library(reshape2)
library(ggplot2)

@

Long chains of amino acids (proteins) or nucleotides (RNA or DNA).

% http://www.uniprot.org/uniprot/Q8I0K7
% https://www.sciencenews.org/article/scorpion-genome-decoded
\vfill

Sample protein sequence:

\texttt{MRKLYCVLLLSAFEFTYMINFGRGQNYWEHPYQKSDVYHP}

\texttt{INEHREHPKEYQYPLHQEHTYQQEDSGEDENTLQHAYPID}

\texttt{HEGAEPAPQEQNLFSSIEIV...}

\end{frame}  

\begin{frame}{Biological sequences}

\begin{figure}
\includegraphics[width=0.55\textwidth]{static_figures/plasmodium.jpg}
\caption{\tiny \textit{Plasmodium falciparum}. Source: \url{http://www.protists.ensembl.org}}
\end{figure}

Protein of \textit{Plasmodium falciparum}:

\texttt{MRKLYCVLLLSAFEFTYMINFGRGQNYWEHPYQKSDVYHP}

\texttt{INEHREHPKEYQYPLHQEHTYQQEDSGEDENTLQHAYPID}

\texttt{HEGAEPAPQEQNLFSSIEIV...}
\end{frame}  



\begin{frame}{Biological sequences}

Signal peptide (red): n-terminal amino acid sequence directing proteins to the endomembrane system and next to extracellular localizations.

\texttt{\textcolor{red}{MRKLYCVLLLSAFEFTYMINFGRGQ}NYWEHPYQKSDVYHP}

\texttt{INEHREHPKEYQYPLHQEHTYQQEDSGEDENTLQHAYPID}

\texttt{HEGAEPAPQEQNLFSSIEIV...}


\end{frame}  

\begin{frame}{Biological sequences}
<<echo = FALSE, message=FALSE,warning=FALSE,fig.height=5>>=
seqs <- read.fasta("./data/seqs.fasta", seqtype = "AA")
seq_df <- lapply(seqs, function(i) i[1L:27]) %>% 
  do.call(rbind, .) %>% 
  cbind(name = sub(">", "", sapply(seqs, function(i) attr(i, "Annot"))), .) %>% 
  data.frame

colnames(seq_df)[-1] <- paste0("P", 1L:27)

mseq_df <- melt(seq_df, id.vars = "name",
     value.name = "aa", variable.name = "pos") %>% 
  mutate(aa = factor(aa, levels = a()[-1])) %>% 
  mutate(pos = as.numeric(sub("P", "", pos))) %>% 
  mutate(name = factor(gsub(" ", "\n", name))) %>% 
  mutate(name = relevel(name, "Plasmodium\nfalciparum")) %>% 
  mutate(sig = pos == "25") %>% 
  inner_join(data.frame(aa = a()[-1], hydr = aaprop[2, ]))

ggplot(mseq_df, aes(x = pos, y = name, label = aa, color = aa)) +
  #geom_point(aes(x = pos, y = name, shape = sig), size = 7, color = "red", stroke = 1) +
  geom_text(size = 5) +
  scale_x_continuous("Position") + 
  scale_y_discrete("") +
  scale_color_discrete(guide = FALSE) +
  #scale_shape_manual(guide = FALSE, values = c(NA, 21)) +
  geom_vline(xintercept = 24.5, color = "red") +
  my_theme +
  theme(axis.text = element_text(size=12),
        panel.grid.major = element_line(color="NA", 
                                        linetype = "dashed", size = 0.5))
@
\end{frame}  


\begin{frame}{n-grams}
n-grams (k-tuples) are vectors of $n$ characters derived from input sequence(s). 


\end{frame}


\begin{frame}{Properites of amino acids}
<<echo = FALSE, message=FALSE,warning=FALSE,fig.height=4>>=
hydr_df <- data.frame(aa = a()[-1], hydr = aaprop[2, ])

ggplot(hydr_df, aes(x = aa, y = hydr)) + 
  geom_bar(stat = "identity") + 
  scale_x_discrete("Amino acid") +
  scale_y_continuous("Hydrophobicity") +
  my_theme
@

Amino acids may be described using their physicochemical properites.

\end{frame}


\begin{frame}{Biological sequences}

<<echo = FALSE, message=FALSE,warning=FALSE,fig.height=5>>=
ggplot(filter(mseq_df, pos != 1), aes(x = pos, y = hydr, group = name)) +
  geom_point() +
  geom_line() +
  #geom_point(aes(x = pos, y = name, shape = sig), size = 7, color = "red", stroke = 1) +
  scale_x_continuous("Position") + 
  scale_y_continuous("Hydrophobicity") +
  scale_color_discrete(guide = FALSE) +
  geom_smooth() +
  my_theme +
  facet_wrap(~ name)
@


\end{frame}  



\begin{document}
  \bibliographystyle{apalike}
  \bibliography{amyloids}
\end{frame}

  
\end{document}